Focused-Launch: Refactor Plan Summary Items rendering

StefanNieuwenhuis opened the pull request:
Currently, the items are rendered separately and in order to make the code DRY-er and more maintainable it makes more sense to collect them in an array and render them in a .map().

#### Testing instructions
* checkout this branch
* run `yarn` && `yarn start` in a terminal;
* go to the block editor, append `?flags=create/focused-launch-flow-calypso` to the URL, and refresh the page;
* compare the plans rendering behavior with production. They should match!

Fixes #47994

StefanNieuwenhuis committed:
Focused-Launch: Refactor Plan Summary Items rendering

Currently the items are rendered seperately and in order to make the code DRY-er and more maintainable it makes more sense to collect them in an array and render them in a .map().

alshakero commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -369,80 +375,44 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
+									plan && (
+										<FocusedLaunchSummaryItem
+											isLoading={ ! defaultFreePlan || ! defaultPaidPlan }
+											onClick={ () => plan && setPlan( plan ) }
+											isSelected={ plan && isPlanSelected( plan ) }
```
We know that `plan` is not null at this point.

StefanNieuwenhuis replied:
You're absolutely right and the check on `plan` is redundant, hence I removed it (and the other's as well).

alshakero commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -369,80 +375,44 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
+												}
+												badgeText={ plan?.isPopular ? __( 'Popular', __i18n_text_domain__ ) : '' }
+											/>
+											{ plan.isFree ? (
+												<TrailingContentSide
+													nodeType={ hasPaidDomain || selectedPaidDomain ? 'WARNING' : 'PRICE' }
+												>
+													{ hasPaidDomain || selectedPaidDomain
+														? __( 'Not available with your domain selection', __i18n_text_domain__ )
+														: __( 'Free', __i18n_text_domain__ ) }
+												</TrailingContentSide>
+											) : (
+												<TrailingContentSide nodeType="PRICE">
+													<span>{ plan && planPrices[ plan?.storeSlug ] }</span>
+													<span>
+														{
+															// translators: /mo is short for "per-month"
+															__( '/mo', __i18n_text_domain__ )
+														}
```
I really think this is now less readable than before. 3-level conditionals that are pretty hard to wrap my head around.

alshakero reviewed (changes requested):
While this is indeed DRYer, I think it made things too complicated for little value. 

I agree the original state isn't the best, it feels unclean. But I have to say that this isn't the way to go. The two items are not similar enough to be looped over and this created multi-level ternary statements. 

It's totally up to you, but maybe a better way would be to figure all this out in pure TypeScript before the JSX starts and then loop over them without any conditions in the JSX.

alshakero commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -369,80 +375,44 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
+											<LeadingContentSide
+												label={
+													/* translators: %s is WordPress.com plan name (eg: Premium Plan) */
+													sprintf( __( '%s Plan', __i18n_text_domain__ ), plan?.title ?? '' )
```
We know `plan` isn't falsey.

alshakero commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -369,80 +375,44 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
+													/* translators: %s is WordPress.com plan name (eg: Premium Plan) */
+													sprintf( __( '%s Plan', __i18n_text_domain__ ), plan?.title ?? '' )
+												}
+												badgeText={ plan?.isPopular ? __( 'Popular', __i18n_text_domain__ ) : '' }
```
We know `plan` isn't falsey.

alshakero commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -369,80 +375,44 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
+												</TrailingContentSide>
+											) : (
+												<TrailingContentSide nodeType="PRICE">
+													<span>{ plan && planPrices[ plan?.storeSlug ] }</span>
```
We know `plan` isn't falsey.

StefanNieuwenhuis committed:
Focused-Launch: Remove redundant checks on plan

alshakero commented on packages/launch/src/focused-launch/summary/index.tsx:
```diff
@@ -369,80 +375,44 @@ const PlanStep: React.FunctionComponent< PlanStepProps > = ( {
+							{ allAvailablePlans.map(
+								( plan ) =>
+									plan && (
+										<FocusedLaunchSummaryItem
```
I think these need a `key` since they're in a loop now.

alshakero reviewed (approved):
LGTM.

StefanNieuwenhuis committed:
Focused-Launch: Add key to FocusedLaunchSummaryItem

Since FocusedLaunchSummaryItem are rendered w/ a .map() we have to add a key prop to it.

StefanNieuwenhuis committed:
Focused-Launch: Replace generated key with storeSlug

alshakero reviewed (approved):
🚢 It
